src/util/basic_types.hpp

```diff
@@ -41,9 +95,103 @@ template<> struct truncate_check_impl<int64_t, size_t, 0> {
   }
 };
 
+// Convenience macro for declaring types allocated via shared_ptr.
+#define DECL_STRUCT(X) \
+  struct X; \
+  using X ## _p = shared_ptr<X>; \
+  bool operator==(shared_ptr<X>, shared_ptr<X>); \
+
 // For integer values, returns ceil(n / m);
 template<typename T> T ceil_divide(T n, T m) {
   return (n + (m-1)) / m;
 }
 
+using i8 = int8_t;
+using u8 = uint8_t;
+using i16 = int16_t;
+using u16 = uint16_t;
+using i32 = int32_t;
+using u32 = uint32_t;
+using f32 = float;
+using i64 = int64_t;
+using u64 = uint64_t;
+using f64 = double;
+
+template<typename T> i64 len(const std::vector<T>& x) {
+  return static_cast<i64>(x.size());
+}
+
+inline i64 len(const std::string& x) {
+  return static_cast<i64>(x.length());
+}
+
+template<typename T> i64 len(const std::unordered_set<T>& x) {
+  return static_cast<i64>(x.size());
+}
+
+template<typename T> i64 len(const std::set<T>& x) {
+  return static_cast<i64>(x.size());
+}
+
+template<typename T> bool all_distinct(const std::vector<T>& v) {
+  std::unordered_set<T> vs;
+  for (const auto& vi : v) {
+    vs.insert(vi);
+  }
+  return (len(vs) == len(v));
+}
+
+template<typename T> optional<T> vector_max(const std::vector<T>& v) {
+  if (v.size() == 0) {
+    return NONE<T>();
+  }
+
+  T ret = v[0];
+  for (i64 i = 1; i < len(v); i++) {
+    ret = std::max<T>(ret, v[i]);
+  }
+
+  return SOME(ret);
+}
+
+template<typename T> T product(const std::vector<T>& x) {
+  T ret = 1;
+  for (i64 i = 0; i < len(x); i++) {
+    ret *= x[i];
+  }
+  return ret;
+}
+
+std::vector<i64> contiguous_strides(const std::vector<i64>& shape);
+
+template<typename T> T& at(std::vector<T>& v, i64 i) {
+  ASSERT_LT(i, len(v));
+  return v[i];
+}
+
+template<typename T> const T& at(const std::vector<T>& v, i64 i) {
+  ASSERT_LT(i, len(v));
+  return v[i];
+}
+
+template<typename T, typename U> std::ostream& operator<<(
+  std::ostream& os, const std::pair<T, U>& x) {
+
+  os << "(" << x.first << ", " << x.second << ")";
+  return os;
+}
+
+namespace std {
+  template<typename T, typename U> struct hash<pair<T, U>> {
```

Comment:
Not sure it's valid in C++ to template-specialize hash for built in types. See https://stackoverflow.com/a/20602159. Does the C++ standard not define hash<pair<T,U>> for us, where T and U are hashable themselves?